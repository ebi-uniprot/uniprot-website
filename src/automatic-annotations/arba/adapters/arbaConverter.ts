import { AAModel } from '../../shared/model';

export type ARBAAPIModel = AAModel & {
  // TODO: change that in the model (backend)
  uniRuleId: string;
};

export type ARBAUIModel = ARBAAPIModel & {
  // any addition/change by the converter
};

const uniRuleConverter = (data: ARBAAPIModel): ARBAUIModel => ({
  ...data,
});

export default uniRuleConverter;
