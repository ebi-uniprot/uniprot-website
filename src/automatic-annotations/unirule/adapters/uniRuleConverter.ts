import { AAModel } from '../../shared/model';

export type UniRuleAPIModel = AAModel & {
  uniRuleId: string;
};

export type UniRuleUIModel = UniRuleAPIModel & {
  // any addition/change by the converter
};

const uniRuleConverter = (data: UniRuleAPIModel): UniRuleUIModel => ({
  ...data,
});

export default uniRuleConverter;
