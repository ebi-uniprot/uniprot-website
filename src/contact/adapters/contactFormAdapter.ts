import { AxiosRequestConfig } from 'axios';
import { debounce } from 'lodash-es';
import {
  Dispatch,
  FormEventHandler,
  SetStateAction,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { useLocation, useNavigate } from 'react-router';

import { Location, LocationToPath } from '../../app/config/urls';
import { addMessage } from '../../messages/state/messagesActions';
import {
  MessageFormat,
  MessageLevel,
} from '../../messages/types/messagesTypes';
import useDataApi from '../../shared/hooks/useDataApi';
import useMessagesDispatch from '../../shared/hooks/useMessagesDispatch';
import { stringifyUrl } from '../../shared/utils/url';
import apiUrls from '../config/apiUrls';

export const modifyFormData = (formData: FormData, token: string) => {
  const output = new FormData();
  output.set('token', token);
  output.set('email', formData.get('email') || '');
  output.set('subject', `[uuw] ${formData.get('subject') || ''}`);
  output.set('requiredForRobots', formData.get('requiredForRobots') || '');
  const message = `${formData.get('message') || ''}

Name: ${formData.get('name')}

--------------- prefilled context details ---------------

${formData.get('context') || ''}`;
  output.set('message', message);
  return output;
};

const DEBOUNCE_TIME = 1_000;

export type Suggestion =
  | 'update'
  | 'PDB'
  | 'buy'
  | 'blast'
  | 'align'
  | 'id mapping'
  | 'peptide search'
  | 'not English';
const entryUpdateRegExp = /entry update request/i;
const pdbRegExp = /(structure)|(pdb)/i;
const buyRegExp = /(buy)|(purchas)|(ship)|(cost)|(quote)/i;
const blastRegExp = /blast/i;
const alignRegExp = /(clustal)|(align)/i;
const idMappingRegExp = /(map)/i;
const peptideSearchRegExp = /(peptide search)|(search peptide)/i;
// anything not outside the printable ASCII range
// eslint-disable-next-line no-control-regex
let nonLatin = /[^\x00-\x7F]/;
try {
  // any Unicode letter minus characters whose Script Extensions include Latin
  // α: true, 東: true, t: false, é: false
  nonLatin = /[\p{L}--\p{scx=Latin}]/v;
} catch {
  /* */
}

export const getSuggestion = debounce(
  (
    callback: Dispatch<SetStateAction<Suggestion[] | undefined>>,
    subject: string = '',
    message: string = ''
  ): void => {
    const suggestions: Suggestion[] = [];
    if (entryUpdateRegExp.test(subject)) {
      // A bit specific, only test the subject as it would have been added by
      // the link from entry page to suggest an update
      suggestions.push('update');
    }
    if (pdbRegExp.test(subject) || pdbRegExp.test(message)) {
      suggestions.push('PDB');
    }
    if (buyRegExp.test(subject) || buyRegExp.test(message)) {
      suggestions.push('buy');
    }
    if (blastRegExp.test(subject) || blastRegExp.test(message)) {
      suggestions.push('blast');
    }
    if (alignRegExp.test(subject) || alignRegExp.test(message)) {
      suggestions.push('align');
    }
    if (idMappingRegExp.test(subject) || idMappingRegExp.test(message)) {
      suggestions.push('id mapping');
    }
    if (
      peptideSearchRegExp.test(subject) ||
      peptideSearchRegExp.test(message)
    ) {
      suggestions.push('peptide search');
    }
    if (nonLatin.test(message)) {
      suggestions.push('not English');
    }
    if (!suggestions.length) {
      callback(undefined);
    } else {
      callback(suggestions);
    }
  },
  DEBOUNCE_TIME
);

export type UseFormLogicReturnType = {
  sending: boolean;
  suggestions: undefined | Suggestion[];
  handleSubmit: FormEventHandler<HTMLFormElement>;
  handleChange: FormEventHandler<HTMLInputElement | HTMLTextAreaElement>;
};

export const useFormLogic = (referrer?: string): UseFormLogicReturnType => {
  const dispatch = useMessagesDispatch();
  const navigate = useNavigate();
  const { state } = useLocation();

  const [formData, setFormData] = useState<undefined | FormData>();
  const [suggestions, setSuggestions] = useState<undefined | Suggestion[]>(
    undefined
  );

  const subject = formData?.get('subject') as undefined | string;

  const tokenData = useDataApi<{ token: string }>(
    subject ? stringifyUrl(apiUrls.token, { key: subject }) : null
  );
  const token = tokenData.data?.token;

  // Generate configuration to POST data to the contact send endpoint
  const postData: undefined | AxiosRequestConfig<FormData> = useMemo(() => {
    if (!formData || !token) {
      return;
    }
    const modifiedFormData = modifyFormData(formData, token);

    return {
      method: 'POST',
      data: modifiedFormData,
    };
  }, [formData, token]);

  const sendData = useDataApi(postData ? apiUrls.send : null, postData);

  const loading = tokenData.loading || sendData.loading;
  const error = tokenData.error || sendData.error;

  // Error behaviour
  useEffect(() => {
    if (error) {
      // Pop up message for the user
      dispatch(
        addMessage({
          format: MessageFormat.POP_UP,
          level: MessageLevel.FAILURE,
          content: 'Error while sending your message. Please try again later.',
        })
      );
    }
  }, [error, dispatch]);

  // Success behaviour
  useEffect(() => {
    if (sendData.data) {
      // Pop up message for the user
      dispatch(
        addMessage({
          format: MessageFormat.POP_UP,
          level: MessageLevel.SUCCESS,
          content: 'Your message has been succesfully sent to our team.',
        })
      );
      // Clear the current form (just in case)
      document.querySelector<HTMLFormElement>('main form')?.reset();
      // Navigate the user to previous page, or to the homepage if not possible
      navigate(referrer || LocationToPath[Location.Home]);
    }
  }, [sendData.data, dispatch, navigate, referrer]);

  const previousLoading = useRef(loading);
  // Reset custom hook form data (to retry sending if needed)
  useEffect(() => {
    if (previousLoading.current && !loading) {
      setFormData(undefined);
    }
    previousLoading.current = loading;
  }, [loading]);

  const handleSubmit: FormEventHandler<HTMLFormElement> = useCallback(
    (event) => {
      const form = event.target;
      event.preventDefault();
      if (!(form instanceof HTMLFormElement) || loading) {
        return; // Shouldn't happen, but just in case
      }
      setFormData(new FormData(form));
    },
    [loading]
  );

  // on unmount, cancel any pending suggestion calculation
  useEffect(() => getSuggestion.cancel, []);

  const handleChange: FormEventHandler<
    HTMLInputElement | HTMLTextAreaElement
  > = (event) => {
    const element = event.target;
    if (
      !(
        element instanceof HTMLInputElement ||
        element instanceof HTMLTextAreaElement
      ) ||
      loading
    ) {
      return;
    }
    getSuggestion(
      setSuggestions,
      element.form?.subject?.value,
      element.form?.message?.value
    );
    navigate('.', {
      replace: true,
      state: {
        ...state,
        formValues: {
          ...(state?.formValues || {}),
          [element.name]: element.value,
        },
      },
    });
  };

  return { sending: loading, suggestions, handleSubmit, handleChange };
};
