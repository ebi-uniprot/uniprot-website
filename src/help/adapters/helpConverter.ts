import { SetRequired } from 'type-fest';
import { SearchResults } from '../../shared/types/results';

export type HelpAPIModel = {
  id: string;
  title: string;
  lastModified: string;
  releaseDate?: string; // specific to news / release notes
  categories?: string[];

  type: 'help' | 'releaseNotes';
  content?: string;

  matches?: {
    // key is one of the keys of the top level object (except "matches" itself)
    [key in Exclude<keyof HelpAPIModel, 'matches'>]?: string[];
  };
};

export type HelpUIModel = HelpAPIModel & {
  // any addition/change by the converter
};

export type HelpSearchResponse = SearchResults<HelpAPIModel>;

export type HelpEntryResponse = SetRequired<
  Omit<HelpAPIModel, 'matches'>,
  'content'
>;

const helpConverter = (data: HelpAPIModel): HelpUIModel => ({
  ...data,
});

export default helpConverter;
