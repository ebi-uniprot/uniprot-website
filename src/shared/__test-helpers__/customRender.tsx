/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
