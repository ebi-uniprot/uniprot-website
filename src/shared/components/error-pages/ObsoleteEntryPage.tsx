import { FC, Fragment } from 'react';
import { Message } from 'franklin-sites';
import { Link } from 'react-router-dom';

import listFormat from '../../utils/listFormat';

import {
  Location,
  LocationToPath,
  getEntryPath,
} from '../../../app/config/urls';
import { Namespace } from '../../types/namespaces';

import {
  InactiveEntryReason,
  InactiveReasonType,
} from '../../../uniprotkb/adapters/uniProtkbConverter';

import ArtWork from './svgs/obsolete-entry.svg';

import './styles/error-pages.scss';

const DeletedEntryMessage: FC<{ accession: string }> = ({ accession }) => (
  <Message level="info">
    <h4>This entry is obsolete</h4>
    <p data-testid="deleted-entry-message">
      The protein sequence for this entry is available in{' '}
      <Link
        to={{
          pathname: LocationToPath[Location.UniParcResults],
          search: `query=${accession}`,
        }}
      >
        UniParc
      </Link>
      . For previous versions of this entry, please look at its{' '}
      <Link
        to={{
          pathname: getEntryPath(Namespace.uniprotkb, accession),
          // TODO: actually implement this in the website
          search: 'version=*',
        }}
      >
        history
      </Link>
      .
    </p>
  </Message>
);

const DemergedEntryMessage: FC<{
  accession: string;
  demergedTo?: string[];
}> = ({ accession, demergedTo }) => (
  <Message level="info">
    <h4>This entry is obsolete</h4>
    {demergedTo?.length && (
      <p data-testid="demerged-entry-message">
        It can now be found as secondary accession in{' '}
        {demergedTo.map((newEntry, index) => (
          <Fragment key={newEntry}>
            {listFormat(index, demergedTo)}
            <Link to={getEntryPath(Namespace.uniprotkb, newEntry)}>
              {newEntry}
            </Link>
          </Fragment>
        ))}
        . [
        <Link
          to={{
            pathname: LocationToPath[Location.UniProtKBResults],
            search: `query=replaces:${accession}`,
          }}
        >
          List
        </Link>
        ]
      </p>
    )}
    <p>
      The protein sequence for this entry is available in{' '}
      <Link
        to={{
          pathname: LocationToPath[Location.UniParcResults],
          search: `query=${accession}`,
        }}
      >
        UniParc
      </Link>
      . For previous versions of this entry, please look at its{' '}
      <Link
        to={{
          pathname: getEntryPath(Namespace.uniprotkb, accession),
          // TODO: actually implement this in the website
          search: `version=*`,
        }}
      >
        history
      </Link>
      .
    </p>
  </Message>
);

type ObsoleteEntryPageProps = {
  accession: string;
  details: InactiveEntryReason;
};

const ObsoleteEntryPage: FC<ObsoleteEntryPageProps> = ({
  accession,
  details,
}) => (
  <div className="error-page-container">
    <ArtWork className="error-page-container__art-work" />
    {details.inactiveReasonType === InactiveReasonType.DELETED ? (
      <DeletedEntryMessage accession={accession} />
    ) : (
      <DemergedEntryMessage
        accession={accession}
        demergedTo={details.mergeDemergeTo}
      />
    )}
  </div>
);

export default ObsoleteEntryPage;
