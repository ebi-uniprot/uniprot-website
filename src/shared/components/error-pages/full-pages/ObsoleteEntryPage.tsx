import { Fragment } from 'react';
import { Link } from 'react-router-dom';

import listFormat from '../../../utils/listFormat';

import {
  Location,
  LocationToPath,
  getEntryPath,
} from '../../../../app/config/urls';
import { Namespace } from '../../../types/namespaces';
import { TabLocation } from '../../../../uniprotkb/types/entry';

const ObsoleteMessageHeading = ({ release }: { release?: string }) => (
  <h4>This entry is obsolete{release && ` since release ${release}`}</h4>
);

type DeletedEntryMessageProps = {
  accession: string;
  release?: string;
  inHistory?: boolean;
};

export const DeletedEntryMessage = ({
  accession,
  release,
  inHistory,
}: DeletedEntryMessageProps) => (
  <>
    <ObsoleteMessageHeading release={release} />
    <p>
      This entry has now been{' '}
      <strong data-article-id="deleted_accessions">deleted</strong>. The protein
      sequence for this entry is available in{' '}
      <Link
        to={{
          pathname: LocationToPath[Location.UniParcResults],
          search: `query=${accession}`,
        }}
      >
        UniParc
      </Link>
      .
      {inHistory || (
        <>
          {' '}
          For previous versions of this entry, please look at its{' '}
          <Link
            to={getEntryPath(
              Namespace.uniprotkb,
              accession,
              TabLocation.History
            )}
          >
            history
          </Link>
          .
        </>
      )}
    </p>
  </>
);

type MergedEntryMessageProps = {
  accession: string;
  mergedInto: string;
  release?: string;
};

export const MergedEntryMessage = ({
  accession,
  mergedInto,
  release,
}: MergedEntryMessageProps) => (
  <>
    <ObsoleteMessageHeading release={release} />
    <p>
      This entry has now been <strong>merged</strong> into{' '}
      <Link
        to={getEntryPath(Namespace.uniprotkb, mergedInto, TabLocation.Entry)}
      >
        {mergedInto}
      </Link>
      .
    </p>
    <p>
      The protein sequence for this entry is available in{' '}
      <Link
        to={{
          pathname: LocationToPath[Location.UniParcResults],
          search: `query=${accession}`,
        }}
      >
        UniParc
      </Link>
      .
    </p>
  </>
);

type DemergedEntryMessageProps = {
  accession: string;
  demergedTo?: string[];
  release?: string;
  inHistory?: boolean;
};

export const DemergedEntryMessage = ({
  accession,
  demergedTo,
  release,
  inHistory,
}: DemergedEntryMessageProps) => (
  <>
    <ObsoleteMessageHeading release={release} />
    {demergedTo?.length && (
      <p>
        This entry has now been <strong>demerged</strong>. Its accession has
        been set as secondary accession in{' '}
        {demergedTo.map((newEntry, index) => (
          <Fragment key={newEntry}>
            {listFormat(index, demergedTo)}
            <Link
              to={getEntryPath(
                Namespace.uniprotkb,
                newEntry,
                TabLocation.Entry
              )}
            >
              {newEntry}
            </Link>
          </Fragment>
        ))}
        . [
        <Link
          to={{
            pathname: LocationToPath[Location.UniProtKBResults],
            search: `query=sec_acc:${accession}`,
          }}
        >
          List of currently active entries
        </Link>
        ]
      </p>
    )}
    <p>
      The protein sequence for this entry is available in{' '}
      <Link
        to={{
          pathname: LocationToPath[Location.UniParcResults],
          search: `query=${accession}`,
        }}
      >
        UniParc
      </Link>
      .
      {inHistory || (
        <>
          {' '}
          For previous versions of this entry, please look at its{' '}
          <Link
            to={getEntryPath(
              Namespace.uniprotkb,
              accession,
              TabLocation.History
            )}
          >
            history
          </Link>
          .
        </>
      )}
    </p>
  </>
);
