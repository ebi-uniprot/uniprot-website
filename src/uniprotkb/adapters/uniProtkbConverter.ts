import EntrySection from '../types/entrySection';
import convertFunction from './functionConverter';
import { FeatureData } from '../components/protein-data-views/FeaturesView';
import convertDiseaseAndDrugs from './diseaseAndDrugs';
import {
  convertNamesAndTaxonomy,
  NamesAndTaxonomyUIModel,
  ProteinNamesData,
  GeneNamesData,
  OrganismData,
  LineageData,
} from './namesAndTaxonomyConverter';
import convertProteinProcessing from './proteinProcessingConverter';
import convertExpression from './expressionConverter';
import convertSubcellularLocation from './subcellularLocationConverter';
import {
  convertSequence,
  SequenceUIModel,
  EntryAudit,
} from './sequenceConverter';
import { SequenceData } from '../../shared/components/entry/SequenceView';
import { Keyword } from '../utils/KeywordsUtil';
import convertInteraction from './interactionConverter';
import convertFamilyAndDomains from './familyAndDomainsConverter';
import { UIModel } from './sectionConverter';
import convertStructure from './structureConverter';
import convertExternalLinks from './externalLinksConverter';
import Comment, { Xref } from '../types/commentTypes';
import { transfromProperties } from '../utils';
import { Property } from '../types/modelTypes';
import { Reference } from '../types/literatureTypes';

export enum EntryType {
  REVIEWED,
  UNREVIEWED,
  INACTIVE,
  UNIPARC,
}

export const getEntryTypeFromString = (entryTypeString: string) => {
  if (entryTypeString.match(/Inactive/gi)) {
    return EntryType.INACTIVE;
  }
  if (entryTypeString.match(/UniParc/i)) {
    return EntryType.UNIPARC;
  }
  if (entryTypeString.match(/TrEMBL|unreviewed|^tr\|$|^tr$/gi)) {
    return EntryType.UNREVIEWED;
  }
  if (entryTypeString.match(/Swiss-Prot|reviewed|^sp\|$|^sp$/gi)) {
    return EntryType.REVIEWED;
  }
  return undefined;
};

export type UniProtkbAPIModel = {
  proteinDescription?: ProteinNamesData;
  genes?: GeneNamesData;
  organism?: OrganismData;
  virusHosts?: OrganismData[];
  primaryAccession: string;
  secondaryAccessions?: string[];
  uniProtkbId: string;
  proteinExistence: string;
  entryType: string;
  inactiveReason?: InactiveEntryReason;
  comments?: Comment[];
  keywords?: Keyword[];
  features?: FeatureData;
  uniProtKBCrossReferences?: Xref[];
  sequence: SequenceData;
  annotationScore: number;
  entryAudit?: EntryAudit;
  references?: Reference[];
  lineages?: LineageData[];
  // How is that defined? What goes in this?
  extraAttributes?: Record<string, unknown>;
};

export type UniProtkbUIModel = {
  primaryAccession: string;
  uniProtkbId: string;
  proteinExistence: string;
  entryType?: EntryType;
  inactiveReason?: InactiveEntryReason;
  annotationScore: number;
  [EntrySection.Function]: UIModel;
  [EntrySection.NamesAndTaxonomy]: NamesAndTaxonomyUIModel;
  [EntrySection.SubCellularLocation]: UIModel;
  [EntrySection.DiseaseAndDrugs]: UIModel;
  [EntrySection.Phenotypes]: UIModel;
  [EntrySection.ProteinProcessing]: UIModel;
  [EntrySection.Expression]: UIModel;
  [EntrySection.Sequence]: SequenceUIModel;
  [EntrySection.Interaction]: UIModel;
  [EntrySection.Structure]: UIModel;
  [EntrySection.FamilyAndDomains]: UIModel;
  [EntrySection.ExternalLinks]: UIModel;
  references?: Reference[];
  extraAttributes: UniProtkbAPIModel['extraAttributes'];
};

export enum InactiveReasonType {
  MERGED = 'MERGED', // We will never see this as this is followed by a 303 redirect
  DEMERGED = 'DEMERGED',
  DELETED = 'DELETED',
}

export type InactiveEntryReason = {
  inactiveReasonType: InactiveReasonType;
  mergeDemergeTo: string[] | [];
};

export const convertXrefProperties = (xrefs: Xref[]) =>
  xrefs.map((xref) => ({
    ...xref,
    properties: xref.properties
      ? transfromProperties((xref.properties as unknown) as Property[])
      : {},
  }));

const uniProtKbConverter = (data: UniProtkbAPIModel): UniProtkbUIModel => {
  const dataCopy = { ...data };
  if (dataCopy.uniProtKBCrossReferences) {
    dataCopy.uniProtKBCrossReferences = convertXrefProperties(
      dataCopy.uniProtKBCrossReferences
    );
  }

  return {
    primaryAccession: dataCopy.primaryAccession,
    uniProtkbId: dataCopy.uniProtkbId,
    proteinExistence: dataCopy.proteinExistence,
    entryType: getEntryTypeFromString(dataCopy.entryType),
    annotationScore: dataCopy.annotationScore,
    inactiveReason: dataCopy.inactiveReason,
    [EntrySection.Function]: convertFunction(dataCopy),
    [EntrySection.NamesAndTaxonomy]: convertNamesAndTaxonomy(dataCopy),
    [EntrySection.SubCellularLocation]: convertSubcellularLocation(dataCopy),
    [EntrySection.DiseaseAndDrugs]: convertDiseaseAndDrugs(dataCopy),
    [EntrySection.Phenotypes]: convertDiseaseAndDrugs(dataCopy),
    [EntrySection.ProteinProcessing]: convertProteinProcessing(dataCopy),
    [EntrySection.Expression]: convertExpression(dataCopy),
    [EntrySection.Interaction]: convertInteraction(dataCopy),
    [EntrySection.Structure]: convertStructure(dataCopy),
    [EntrySection.Sequence]: convertSequence(dataCopy),
    [EntrySection.FamilyAndDomains]: convertFamilyAndDomains(dataCopy),
    [EntrySection.ExternalLinks]: convertExternalLinks(dataCopy),
    references: dataCopy.references || [],
    extraAttributes: data.extraAttributes,
  };
};

export default uniProtKbConverter;

// TODO this needs to be removed once added
// entrySectionToKeywordCategories.set(EntrySection.Miscellaneous, [
//   KeywordCategory.TECHNICAL_TERM,
// ]);
