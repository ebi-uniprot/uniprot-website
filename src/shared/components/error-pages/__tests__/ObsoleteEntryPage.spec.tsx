import customRender from '../../../__test-helpers__/customRender';

import ObsoleteEntryPage from '../ObsoleteEntryPage';

import { InactiveEntryReason } from '../../../../uniprotkb/adapters/uniProtkbConverter';

import deletedEntryData from '../../../__mocks__/deletedEntryModelData.json';
import demergedEntryData from '../../../__mocks__/demergedEntryModelData.json';

describe('ObsoleteEntryPage component', () => {
  test('should render deleted page', () => {
    const { primaryAccession } = deletedEntryData;
    const { asFragment } = customRender(
      <ObsoleteEntryPage
        accession={primaryAccession}
        details={deletedEntryData.inactiveReason as InactiveEntryReason}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('should render demerged page', () => {
    const { primaryAccession } = demergedEntryData;
    const { asFragment } = customRender(
      <ObsoleteEntryPage
        accession={primaryAccession}
        details={demergedEntryData.inactiveReason as InactiveEntryReason}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
