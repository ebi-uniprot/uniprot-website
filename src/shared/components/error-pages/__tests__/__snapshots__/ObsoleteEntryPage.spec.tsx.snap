// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObsoleteEntryPage component should render deleted page 1`] = `
<DocumentFragment>
  <div
    class="error-page-container"
  >
    <test-file-stub
      classname="error-page-container__art-work"
    />
    <div
      class="message message--info"
      role="status"
    >
      <div
        class="message__side-border"
      />
      <test-file-stub
        height="1.125em"
        width="1.125em"
      />
      <section
        class="message__content"
      >
        <small>
          <h4>
            This entry is obsolete
          </h4>
          <p
            data-testid="deleted-entry-message"
          >
            The protein sequence for this entry is available in 
            <a
              href="/uniparc?query=I8FBX0"
            >
              UniParc
            </a>
            . For previous versions of this entry, please look at its 
            <a
              href="/uniprotkb/I8FBX0/history"
            >
              history
            </a>
            .
          </p>
        </small>
      </section>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ObsoleteEntryPage component should render demerged page 1`] = `
<DocumentFragment>
  <div
    class="error-page-container"
  >
    <test-file-stub
      classname="error-page-container__art-work"
    />
    <div
      class="message message--info"
      role="status"
    >
      <div
        class="message__side-border"
      />
      <test-file-stub
        height="1.125em"
        width="1.125em"
      />
      <section
        class="message__content"
      >
        <small>
          <h4>
            This entry is obsolete
          </h4>
          <p
            data-testid="demerged-entry-message"
          >
            It can now be found as secondary accession in 
            <a
              href="/uniprotkb/P63151/entry"
            >
              P63151
            </a>
             and 
            <a
              href="/uniprotkb/P63150/entry"
            >
              P63150
            </a>
            . [
            <a
              href="/uniprotkb?query=sec_acc:Q00007"
            >
              List
            </a>
            ]
          </p>
          <p>
            The protein sequence for this entry is available in 
            <a
              href="/uniparc?query=Q00007"
            >
              UniParc
            </a>
            . For previous versions of this entry, please look at its 
            <a
              href="/uniprotkb/Q00007/history"
            >
              history
            </a>
            .
          </p>
        </small>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
